ethers-providers/src/transports/quorum.rs

```diff
@@ -190,15 +191,67 @@ impl<T: JsonRpcClientWrapper> QuorumProvider<T> {
 
         let mut numbers = vec![];
         let mut errors = vec![];
+
+        // exit once required quorum weight is achieved
+        let required_weight = self.quorum_weight();
+        let mut weight_achieved = 0;
+
+        // timestamp of when we started
+        let start = SystemTime::now();
+
+        // wait until we've reached quorum
         while !queries.is_empty() {
             let (response, _index, remaining) = future::select_all(queries).await;
             queries = remaining;
             match response {
-                Ok(v) => numbers.push(v),
+                Ok(v) => {
+                    weight_achieved += v.0.weight;
+                    numbers.push(v);
+                    if weight_achieved >= required_weight {
+                        break;
+                    }
+                }
                 Err(e) => errors.push(e),
             }
         }
 
+        let quorum_reached_timestamp = SystemTime::now();
+
+        // Current grace period to wait for remaining requests is set to
+        // how long it took for us to reach quorum
+        let quorum_grace_period = quorum_reached_timestamp.duration_since(start)
+            .unwrap_or_default();
+        let timeout_timestamp = quorum_reached_timestamp.checked_add(quorum_grace_period)
+            .unwrap_or_else(|| SystemTime::now());
+        // try and wait for any remaining requests
+        while !queries.is_empty() {
+            let now = SystemTime::now();
+            // if timeout has been exceeded, break
+            if now >= timeout_timestamp {
+                break;
+            }
+            // how long to wait
+            let timeout_duration = match timeout_timestamp.duration_since(now) {
+                Ok(d) => d,
+                Err(_) => break,
+            };
+            match tokio::time::timeout(timeout_duration, future::select_all(queries)).await {
```

Comment:
just noting that the impl feels a little awkward with the whole "ok we're inside the loop, what's the duration left to when we timeout, ok let's call tokio::time::timeout". I wonder if there's a way to do something instead like "this is the duration left to timeout, ok let's do the loop and wait for these queries" ?

I guess the alternative would be to do something fancy like what's done for `QuorumRequest` and its `Future` impl, but i also understand that feels kinda hardcore. If you're interested in going down that path go for it, otherwise I'd have a slight preference for the single timeout alternative, and if that proves inconvenient I'm also fine with this impl